### PR TITLE
Zephyr v3.3.0 xt

### PR DESCRIPTION
**Pull request by xakep-amatop**
**NOT FOR MERGE
CONTAINS ONLY XEN DOM0 RELATED CHANGES
IT IS ONLY FOR PRELIMINARY REVIEW BY XEN-TROOPS TEAM BEFORE OPENING PULL FOR ZEPHYR**

### CODE REVIEW

**Comment by xakep-amatop** [comment]
fix all Compliance Checks except new typedefs and license MIT

File: drivers/xen/dom0/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2021 EPAM Systems
```
Comment (xakep-amatop) [inline drivers/xen/dom0/CMakeLists.txt:2]: change year here and in other files

File: boards/arm64/xenvm/board.cmake [hunk 2]
```diff
@@ -0,0 +1,14 @@
+set(SUPPORTED_EMU_PLATFORMS qemu)
```
Comment (xakep-amatop) [inline boards/arm64/xenvm/board.cmake:1]: looks like these changes should be dropped too

File: drivers/xen/memory.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
+#include <zephyr/arch/arm64/hypercall.h>
```
Comment (xakep-amatop) [inline drivers/xen/memory.c:1]: add preamble

File: include/zephyr/xen/dom0/zimage.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (xakep-amatop) [inline include/zephyr/xen/dom0/zimage.h]: (note) drop this file here and move it to zephyr zenlib 
@firscity @oleksiimoisieiev FYI

File: include/zephyr/xen/generic.h [hunk 5]
```diff
@@ -32,4 +32,10 @@
 				__ATOMIC_SEQ_CST) ? new : old; \
 })
 
+#define NR_MAGIC_PAGES 4
```
Comment (xakep-amatop) [inline include/zephyr/xen/generic.h]: drop this commit and move changes to zephyr xen-lib
@firscity @oleksiimoisieiev

File: include/zephyr/xen/public/io/xenbus.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
+/* SPDX-License-Identifier: MIT */
```
Comment (xakep-amatop) [inline include/zephyr/xen/public/io/xenbus.h]: drop this commit here and move to xenstore srv app
@firscity @oleksiimoisieiev

File: include/zephyr/xen/dom0/uimage.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
+/*
```
Comment (xakep-amatop) [inline include/zephyr/xen/dom0/uimage.h]: drop the file from zephyr and move this file to xen-lib 
@firscity @oleksiimoisieiev

File: dts/bindings/xen/xen,xen-4.15.yaml [hunk 8]
```diff
@@ -1,11 +1,5 @@
-description: Xen Platform Control Registers
+description: Xen Platform Control Registers v4.15
```
Comment (xakep-amatop) [inline dts/bindings/xen/xen,xen-4.15.yaml:1]: change version in all places to 4.17

File: include/zephyr/xen/public/arch-arm.h [hunk 9]
```diff
@@ -205,15 +205,138 @@ typedef uint64_t xen_pfn_t;
...
-typedef uint64_t xen_ulong_t;
-#define PRI_xen_ulong PRIx64
-
 /*
  * Maximum number of virtual CPUs in legacy multi-processor guests.
  * Only one. All other VCPUS must use VCPUOP_register_vcpu_info.
  */
 #define XEN_LEGACY_MAX_VCPUS 1
 
+typedef uint64_t xen_ulong_t;
```
Comment (xakep-amatop) [inline include/zephyr/xen/public/arch-arm.h:214]: check if these changes from xen 4.17 or not

File: arch/arm64/core/mmu.c [hunk 10]
```diff
@@ -820,7 +820,12 @@ static void enable_mmu_el1(struct arm_mmu_ptables *ptables, unsigned int flags)
 	isb();
 
 	/* Invalidate all data caches before enable them */
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+	/* todo: find more optimal solution */
```
Comment (xakep-amatop) [inline arch/arm64/core/mmu.c:824]: can we merge it into mainline?..

File: include/zephyr/xen/public/domctl.h [hunk 11]
```diff
@@ -0,0 +1,387 @@
+/******************************************************************************
```
Comment (firscity) [inline include/zephyr/xen/public/domctl.h:2]: Add license identifier and fix code style in this file

File: include/zephyr/xen/public/domctl.h [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+#ifndef CONFIG_XEN_INITIAL_DOMAIN
+#error "domctl operations are intended for use by node control tools only"
+#endif
+
+#include "xen.h"
+#include "event_channel.h"
+#include "grant_table.h"
+#include "memory.h"
+
+#define XEN_DOMCTL_INTERFACE_VERSION 0x00000014
```
Comment (firscity) [inline include/zephyr/xen/public/domctl.h]: 0x15 should be in initial commit, need to squash with some upper commit where it was made.

File: include/zephyr/xen/public/io/libxenvchan.h [hunk 13]
```diff
@@ -0,0 +1,88 @@
+/* SPDX-License-Identifier: MIT */
```
Comment (firscity) [inline include/zephyr/xen/public/io/libxenvchan.h]: Move it to zephyr-xenlib

File: drivers/xen/CMakeLists.txt [hunk 14]
```diff
@@ -4,3 +4,4 @@
 zephyr_sources(hvm.c)
 zephyr_sources(events.c)
 zephyr_sources_ifdef(CONFIG_XEN_GRANT_TABLE gnttab.c)
+zephyr_sources(memory.c)
```
Comment (firscity) [inline drivers/xen/CMakeLists.txt:7]: This change should be in separate commit

File: drivers/xen/memory.c [hunk 15]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (firscity) [inline drivers/xen/memory.c:2]: These changes should be in separate commit

File: drivers/xen/memory.c [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+#include <zephyr/kernel.h>
+#include <string.h>
+
+
+int xendom_add_to_physmap(int domid, unsigned long idx,
+		unsigned int space, xen_pfn_t gpfn)
+{
+	struct xen_add_to_physmap xatp;
+
+	memset(&xatp, 0, sizeof(xatp));
```
Comment (firscity) [inline drivers/xen/memory.c]: `xatp = {0}`

File: drivers/xen/memory.c [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+int xendom_add_to_physmap_batch(int domid, int foreign_domid,
+			unsigned int space, unsigned int size,
+			xen_ulong_t *idxs, xen_pfn_t *gpfns,
+			int *errs)
+{
+	struct xen_add_to_physmap_batch xatpb;
+
+	memset(&xatpb, 0, sizeof(xatpb));
```
Comment (firscity) [inline drivers/xen/memory.c]: `xatpb = {0}`

File: drivers/xen/memory.c [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+	return HYPERVISOR_memory_op(XENMEM_remove_from_physmap, &xrfp);
+}
+
+int xendom_populate_physmap(int domid, unsigned int extent_order,
+		unsigned int nr_extents, unsigned int mem_flags,
+		xen_pfn_t *extent_start)
+{
+	struct xen_memory_reservation reservation;
+
+	memset(&reservation, 0, sizeof(reservation));
```
Comment (firscity) [inline drivers/xen/memory.c]: `reservation = {0}`

File: include/zephyr/xen/memory.h [hunk 19]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* TODO: fix headers including and add docs for functions */
```
Comment (firscity) [inline include/zephyr/xen/memory.h]: Fix this todo before upstream

File: drivers/xen/dom0/domctl.c [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <string.h>
+
+int xen_domctl_scheduler_op(int domid, struct xen_domctl_scheduler_op *sched_op)
+{
+	xen_domctl_t domctl;
+
+	memset(&domctl, 0, sizeof(domctl));
```
Comment (firscity) [inline drivers/xen/dom0/domctl.c]: `domctl = {0}` everywhere instead of memset

File: drivers/xen/dom0/domctl.c [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+	domctl.domain = domid;
+	domctl.cmd = XEN_DOMCTL_bind_pt_irq;
+	switch (irq_type) {
+	case PT_IRQ_TYPE_SPI:
+		bind->irq_type = irq_type;
+		bind->machine_irq = machine_irq;
+		bind->u.spi.spi = spi;
+		break;
+	default:
+		/* TODO: check what to do with others */
```
Comment (firscity) [inline drivers/xen/dom0/domctl.c]: Check this todo before upstream

File: include/zephyr/xen/public/domctl.h [hunk 22]
```diff
@@ -37,7 +37,7 @@
 #include "grant_table.h"
 #include "memory.h"
 
-#define XEN_DOMCTL_INTERFACE_VERSION 0x00000014
+#define XEN_DOMCTL_INTERFACE_VERSION 0x00000015
```
Comment (firscity) [inline include/zephyr/xen/public/domctl.h:23]: move to initial commit

File: include/zephyr/xen/public/domctl.h [hunk 23]
```diff
@@ -69,9 +69,11 @@ struct xen_domctl_createdomain {
 #define XEN_DOMCTL_CDF_iommu          (1U<<_XEN_DOMCTL_CDF_iommu)
 #define _XEN_DOMCTL_CDF_nested_virt   6
 #define XEN_DOMCTL_CDF_nested_virt    (1U << _XEN_DOMCTL_CDF_nested_virt)
+/* Should we expose the vPMU to the guest? */
+#define XEN_DOMCTL_CDF_vpmu           (1U << 7)
```
Comment (firscity) [inline include/zephyr/xen/public/domctl.h]: move all changes to public header to initial commit

File: dts/bindings/xen/xen,xen-4.15.yaml [hunk 24]
```diff
@@ -1,11 +1,5 @@
...
 
 compatible: "xen,xen-4.15"
```
Comment (firscity) [inline dts/bindings/xen/xen,xen-4.15.yaml]: Probably we can drop exact version at all, just use "xen,xen"

File: boards/arm64/xenvm/xenvm_gicv3_defconfig [hunk 25]
```diff
@@ -7,7 +7,7 @@ CONFIG_XENVM_USE_GIC_V3=y
 # Enable UART driver
 CONFIG_SERIAL=y
 
-CONFIG_MAX_XLAT_TABLES=10
+CONFIG_MAX_XLAT_TABLES=24
```
Comment (firscity) [inline boards/arm64/xenvm/xenvm_gicv3_defconfig:10]: Same should be done for `xenvm_defconfig`
FYI: @xakep-amatop

File: drivers/xen/dom0/domctl.c [hunk 26]
```diff
@@ -0,0 +1,302 @@
...
+
+static int do_domctl(xen_domctl_t *domctl)
+{
+	domctl->interface_version = XEN_DOMCTL_INTERFACE_VERSION;
+	return HYPERVISOR_domctl(domctl);
+}
+
+int xen_domctl_scheduler_op(int domid, struct xen_domctl_scheduler_op *sched_op)
+{
+	xen_domctl_t domctl = {0};
```
Comment (xakep-amatop) [inline drivers/xen/dom0/domctl.c]: (note) you can use the next initializer here and below:
```
xen_domctl_t domctl = {
	.cmd = XEN_DOMCTL_scheduler_op,
	.domain = domid,
	.u.scheduler_op = *sched_op,
};
```
and rest of fields automatically will be initialized

File: dts/bindings/xen/xen,xen-4.15.yaml (hunk 24 above)
Comment (xakep-amatop) [reply dts/bindings/xen/xen,xen-4.15.yaml]: yes we can

File: drivers/xen/dom0/domctl.c (hunk 26 above)
Comment (firscity) [reply drivers/xen/dom0/domctl.c]: @xakep-amatop thanks for suggestion, done.

**Comment by xakep-amatop** [comment]
note: not needed anymore --- clear changes you can find here: https://github.com/xen-troops/zephyr/pull/37
